File: hardwareUi/gpio/test_nfc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ติดตั้งไลบรารีด้วยคำสั่ง: pip3 install mfrc522 spidev RPi.GPIO

import RPi.GPIO as GPIO
from mfrc522 import SimpleMFRC522
from mfrc522 import MFRC522
import time
import sys

# ปิดการแจ้งเตือน GPIO (ถ้ามีการใช้งานค้างอยู่)
GPIO.setwarnings(False)

def read_tag():
    reader = SimpleMFRC522()
    try:
        print("\n=== โหมดสแกนอ่านค่า (Read Mode) ===")
        print("กรุณานำเหรียญ/บัตร NFC ไปทาบที่ตัวอ่าน (RC522)...")
        id, text = reader.read()
        print("\n✅ หีหรือเหรียญ NFC ที่พบ:")
        print(f"UID (รหัสประจำตัวโรงงาน): {id}")
        print(f"ข้อมูลใน Memory (Text): {text.strip() if text else 'ไม่มีข้อมูล'}")
    except Exception as e:
        print(f"❌ เกิดข้อผิดพลาด: {e}")
    finally:
        GPIO.cleanup()

def write_tag():
    reader = SimpleMFRC522()
    try:
        print("\n=== โหมดเขียนข้อมูล (Write Mode) ===")
        data = input("กรุณาพิมพ์ข้อความที่ต้องการเขียนลงในเหรียญ (ภาษาอังกฤษ/ตัวเลข): ")
        print("กรุณานำเหรียญ/บัตร NFC ไปทาบที่ตัวอ่าน ค้างไว้จนกว่าจะเสร็จ...")
        
        # เขียนข้อมูลลง Block (เฉพาะ NTAG215 หรือ Mifare ที่ไม่ได้ล็อค User Memory)
        # หมายเหตุ: ไม่ใช่การแก้รหัส UID
        reader.write(data)
        
        print(f"\n✅ เขียนข้อมูล '{data}' ลงในเหรียญสำเร็จ!")
    except Exception as e:
        print(f"❌ เกิดข้อผิดพลาด: {e}")
    finally:
        GPIO.cleanup()

def dump_info():
    reader = MFRC522()
    try:
        print("\n=== โหมดสแกน Raw Data (อ่าน UID ชั้นลึก) ===")
        print("กรุณานำเหรียญ/บัตร NFC ไปทาบที่ตัวอ่าน...")
        
        while True:
            (status, TagType) = reader.MFRC522_Request(reader.PICC_REQIDL)
            
            if status == reader.MI_OK:
                (status, uid) = reader.MFRC522_Anticoll()
                if status == reader.MI_OK:
                    uid_hex = "-".join([f"{x:02X}" for x in uid])
                    print(f"\n✅ เจอแท็ก! (Type: {TagType})")
                    print(f"UID (Raw Hex): {uid_hex}")
                    break
            time.sleep(0.1)
    finally:
        GPIO.cleanup()

if __name__ == '__main__':
    print("=========================================")
    print("   โปรแกรมทดสอบเครื่องอ่าน NFC (RC522)    ")
    print("=========================================")
    print("1: อ่านรหัส UID และข้อมูลทั่วไป (Read)")
    print("2: ทดสอบเขียนข้อมูลลง Memory (Write Text)")
    print("3: แสกนรหัส Raw UID แบบดิบ (Raw Hex)")
    print("=========================================")
    
    choice = input("กรุณาเลือกโหมด (1/2/3): ")
    
    if choice == '1':
        read_tag()
    elif choice == '2':
        write_tag()
    elif choice == '3':
        dump_info()
    else:
        print("ไม่พบตัวเลือก ยกเลิกการทำงาน")
        sys.exit(1)

File: hardwareUi/gpio/test_nfc.mjs
import { execSync } from 'child_process';
import readline from 'readline';

console.log("=========================================");
console.log("   โปรแกรมทดสอบเครื่องอ่าน NFC (NodeJS)  ");
console.log("=========================================");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const checkAndImport = async () => {
    try {
        console.log("กำลังเชื่อมต่อบอร์ด MFRC522...");
        const { default: Mfrc522Lib } = await import('mfrc522-rpi');
        return new Mfrc522Lib();
    } catch (e) {
        console.error("❌ ไม่สามารถโหลด mfrc522-rpi ได้ กรุณาเช็คว่ารันบน RPi และลง package ครบแล้ว", e.message);
        process.exit(1);
    }
};

const readTag = (mfrc522) => {
    console.log("\n[โหมดอ่าน UID (Read)]");
    console.log("กรุณานำเหรียญ/บัตร NFC ไปทาบที่ตัวอ่าน (RC522) ภายใน 10 วินาที...");

    // ตั้งเวลา 10 วิ หมดเวลาจะตัดจบ
    let elapsed = 0;
    const interval = setInterval(() => {
        const found = mfrc522.findCard();
        if (found?.status) {
            const uidResult = mfrc522.getUid();
            if (uidResult?.status) {
                const uid = uidResult.data
                    .slice(0, 4)
                    .map((b) => b.toString(16).padStart(2, '0'))
                    .join('')
                    .toUpperCase();
                console.log(`\n✅ เจอแท็ก! UID: ${uid}`);
                clearInterval(interval);
                process.exit(0);
            }
        }
        elapsed += 0.5;
        if (elapsed >= 10) {
            console.log("⏳ หมดเวลาแสกน ไม่พบเหรียญ");
            clearInterval(interval);
            process.exit(0);
        }
    }, 500);
};

const writeTag = (mfrc522) => {
    console.log("\n❌ [โหมดเขียนข้อมูล]");
    console.log("ตัวไลบรารี mfrc522-rpi ไม่รองรับการเขียนข้อมูล หรือเหรียญ NTAG215 จำนวนมากปฏิเสธการแก้ UID");
    console.log("แนะนำให้ใช้ โหมดอ่าน (กด 1) เพื่อดึงรหัสโรงงานไปเซฟจะเสถียรที่สุดครับ!");
    process.exit(0);
};

const start = async () => {
    const mfrc522 = await checkAndImport();

    console.log("1: อ่านรหัส UID จากเหรียญ (Read)");
    console.log("2: ทดสอบเขียนข้อมูล (Write)");
    console.log("=========================================");

    rl.question("กรุณาเลือกโหมด (1/2): ", (choice) => {
        if (choice === '1') {
            readTag(mfrc522);
        } else if (choice === '2') {
            writeTag(mfrc522);
        } else {
            console.log("ไม่พบตัวเลือก ยกเลิก");
            process.exit(1);
        }
    });
};

start();
